README.md
<div align="center">
  <img src="https://raw.githubusercontent.com/OpenStreamDeck/StreamDeckSharp/master/doc/images/banner/StreamDeckSharpBanner_150px.png">
</div>

-----------------

**StreamDeckSharp is a simple (unofficial) .NET interface for the [Elgato Stream Deck](https://www.elgato.com/en/gaming/stream-deck)**

[![license](https://img.shields.io/github/license/OpenStreamDeck/StreamDeckSharp.svg)](https://github.com/OpenStreamDeck/StreamDeckSharp/blob/master/LICENSE.md)
[![GitHub release](https://img.shields.io/github/release/OpenStreamDeck/StreamDeckSharp.svg)](https://github.com/OpenStreamDeck/StreamDeckSharp/releases)
[![Nuget version](https://img.shields.io/nuget/v/streamdecksharp.svg)](https://www.nuget.org/packages/StreamDeckSharp/)

#### [Recent Changes](CHANGELOG.md)

## Changes in v4.0.0-preview
A preview version (v4.0.0-preview) is currently available on nuget. At the moment there isn't a migration guide, but Jon Skeet has a project that show a the changes he had to make: https://github.com/OpenMacroBoard/StreamDeckSharp/issues/45#issuecomment-1079933751
Alternativly you can also take a look at the changes of the example projects: [Changes example projects](https://github.com/OpenMacroBoard/OpenMacroBoard.ExampleCollection/compare/1104c4f...3203bff)

## Quickstart _(TL;DR)_
***At the moment only Windows is "officially" supported.***  
*Need help with Linux or macOS? → Wiki: [Supported Platforms](https://github.com/OpenMacroBoard/StreamDeckSharp/wiki/Supported-Platforms-and-Devices)*

1. Add StreamDeckSharp reference (via nuget or download latest release)
2. Add a using directive for StreamDeckSharp: `using StreamDeckSharp;`

I want to...              | Code (C#)
------------------------- | ---------------------------------------------------------
create a device reference | `var deck = StreamDeck.OpenDevice();`  
set the brightness        | `deck.SetBrightness(50);`
create bitmap for key     | `var bitmap = KeyBitmap.Create.FromFile("icon.png")`
set key image             | `deck.SetKeyBitmap(keyId, bitmap)`
clear key image           | `deck.ClearKey(keyId)`
process key events        | `deck.KeyStateChanged += KeyHandler;`

**Make sure to dispose the device reference correctly** _(use `using` whenever possible)_

## Examples
If you want to see some examples take a look at the [example projects](https://github.com/OpenMacroBoard/StreamDeckSharp.ExampleCollection).  
Here is a short example called "Austria". Copy the code and start hacking :wink:

```C#
using System;
using OpenMacroBoard.SDK;
using StreamDeckSharp;

namespace StreamDeckSharp.Examples.Austria
{
    class Program
    {
        static void Main(string[] args)
        {
            //This example is designed for the 5x3 (original) Stream Deck.

            //Create some color we use later to draw the flag of austria
            var red = KeyBitmap.Create.FromRgb(237, 41, 57);
            var white = KeyBitmap.Create.FromRgb(255, 255, 255);
            var rowColors = new KeyBitmap[] { red, white, red };

            //Open the Stream Deck device
            using (var deck = StreamDeck.OpenDevice())
            {
                deck.SetBrightness(100);

                //Send the bitmap informaton to the device
                for (int i = 0; i < deck.Keys.Count; i++)
                    deck.SetKeyBitmap(i, rowColors[i / 5]);

                Console.ReadKey();
            }
        }
    }
}
```

Here is what the "Rainbow" example looks like after pressing some keys

![Rainbow example photo](doc/images/rainbow_example.png?raw=true "Rainbow demo after pressing some keys")

### Play games on a StreamDeck
For example minesweeper (take a look at the [example projects](https://github.com/OpenMacroBoard/StreamDeckSharp.ExampleCollection) if you are interested in that)

<img src="doc/images/minesweeper.jpg?raw=true" width="500" />

### You can even play videos on a StreamDeck
Here is a short demo, playing a video on a stream deck device.

[![Demo video of the example](https://i.imgur.com/8tlkaIg.png)](http://www.youtube.com/watch?v=tNwUG0sPmKw)  
_*The glitches you can see are already fixed._

More about that in the Wiki: [Play video on StreamDeck](https://github.com/OpenStreamDeck/StreamDeckSharp/wiki/Play-video-on-StreamDeck)

---
 
###### This project is not related to *Elgato Systems GmbH* in any way

---


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.
I'm trying to keep it up to date, but I'm a lazy bastard - when in doubt - check out the commit log ;-)

## [v5.0.0-preview] - 2023-08-01
Again another preview because I'm pretty busy right now.
- Restructure how the HID communication is abstracted internally. This is useful for consumers when Elgato releases new hardware that can be controlled with the same software. The user of the library can now register new devices with existing "HID drivers" (for example if a Stream Deck MK.3 is released)
- `KeyBitmap.FromBgr24Array` was moved to `KeyBitmap.Create.FromBgr24Array`
- More HID Verification Tests
- Tested with real hardware
  - Stream Deck XL (FW v1.01.000)
  - Stream Deck Mini (FW v2.03.001)
  - Stream Deck (FW v1.0.191203)
  - Stream Deck MK.2 (FW v1.01.001)
    - Saw visual glitches in ImageGlitchTest.Rainbow without throttling
    - Added throttling to about 1,5 MB/s
    - Example video frame rate looked still ok with throttling

## [v4.0.0-preview] - 2022-11-18
A preview version (v4.0.0-preview) is currently available on nuget. At the moment there isn't a migration guide, but Jon Skeet has a project that show a the changes he had to make: https://github.com/OpenMacroBoard/StreamDeckSharp/issues/45#issuecomment-1079933751
Alternatively you can also take a look at the changes of the example projects: [Changes example projects](https://github.com/OpenMacroBoard/OpenMacroBoard.ExampleCollection/compare/1104c4f...3203bff)

## [3.2.0] - 2021-09-04
- Support for Stream Deck MK.2 (by jdahlblom - https://github.com/OpenMacroBoard/StreamDeckSharp/pull/36)
- Remove old cooldown mechanism and replace with a new throttle mechanism that limits the write speed
- Fix a few reported gaps between buttons (again)

## [3.0.0] - 2021-03-05
- Added HidSharp Source to SteamDeckSharp, because the original HidSharp does not have a strong name we sign it here and ship it with StreamDeckSharp. This is required so we can ship new versions of StreamDeckSharp that reference an already existing version of OpenStreamDeck.SDK

## [2.2.0] - 2020-08-24
- Improve accuracy of the reported gaps between buttons (is used to draw "fullscreen" for example)

## [2.1.0] - 2020-08-22
- Added `StreamDeck.CreateDeviceListener()` that allows to keep track of stream decks between reconnects and provides events to react to connection changes.

## [2.0.0] - 2020-06-05
- Drop .NET 4 support. In theory one could still port back the code to .NET 4 but it's not worth it to still maintain that. netstandard20 is the only target from now on.

## [1.1.0] - 2020-03-14
- A lot of code cleanup (+ static analyzers to enforce rules)
- Added locks around hidstream access to fix some glitches
- Switch to semantic(-ish) versioning
- Fix alignment issue with keys that weren't 72px but 96px

## [0.3.5] - 2019-11-28
### Changed
  - Prepare for netstandard2 release
  - Switch to HidSharp (because of netstandard support)
  - Change the stream deck Xl jpg encoder (the old one was from WPF, which is not supported on netstandard)
  - Switch to new csproj format and embedded nuspec
  - Remove old build stuff (switched to nuke.build)
  - New icon for the nuget package

### Added
  - Support for Stream Deck Rev2 :tada: (Thanks to patrick-dmxc for reporting that)

## [0.3.4] - 2019-06-22
### Added
  - Stream Deck XL now correctly
      - reports firmware version
      - reports serial number
      - shows the logo
      - sets the brightness

### Changed
  - It's now possible to open devices without caching (not recommended)
  - Lower key cooldown for Mini and XL variant, because they don't suffer
    from images glitches like the classic stream deck. This improves the
    framerate for cached handles (try the video example ^^)
  - `IHardware` now has a `DeviceName` property
  - Cleaned up a lot of things

## [0.3.2] - 2019-06-19
### Added
  - Support for Stream Deck XL :tada:

## [0.3.1] - 2019-01-26
### Added
  - Implement `GridKeyPositionCollection` to make dealing with keyboard layouts simpler
  - New `IStreamDeckBoard` for easier access to `GridKeyPositionCollection` Keys
    (Use pattern matching on `IMacroBoard` to get `GridKeyPositionCollection`)
  - Methods to get serialnumber and firmware version from `IStreamDeckBoard`

### Changed
  - Change nuget package to license expression

## [0.2.0] - 2018-08-25 - *OpenMacroBoard* :tada:
Elgato Systems released the *"Stream Deck Mini"* a few weeks ago so I decided to refactor
`StreamDeckSharp` and use a more generic approach that makes it possible to implement
many different macro boards (with LCD buttons) - even ones with buttons that are not placed in a grid layout (like keyboards).

Because of this changes I decided to split `StreamDeckSharp` into a generic part called `OpenMacroBoard.SDK`
and the StreamDeck specific part still called `StreamDeckSharp` and to change the organization name from `OpenStreamDeck` to `OpenMacroBoard` to reflect the generic nature.

### Added
  - `OpenMacroBoard.VirtualBoard` to allow developing software for macro boards (eg. Stream Deck) without real hardware.
  - A bunch of unit tests
  - `IKeyBitmapDataAccess` to retrieve raw RGB data from KeyBitmaps
  - `IKeyPositionCollection` to allow for complex key layouts.
  - `DrawFullScreenExtension` is now part of the library (instead of just an example project)
  - More unit tests
  - Support for different LCD key resolutions
### Changed
  - Interface name `IStreamDeck` to `IMacroBoard`
  - KeyId order for the stream deck changed to be more intuitive
    _(left-to-right and top-to-bottom)_

## [0.1.10] - 2018-04-30
### Added
  - StreamDeck.EnumerateDevices to discover all connected StreamDeck devices.

## [0.1.9] - 2018-02-09
### Added

src/StreamDeckSharp/Hardware.cs
using OpenMacroBoard.SDK;
using StreamDeckSharp.Internals;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;

using static StreamDeckSharp.UsbConstants;

namespace StreamDeckSharp
{
    /// <summary>
    /// Details about different StreamDeck Hardware
    /// </summary>
    public static class Hardware
    {
        private static readonly ConcurrentDictionary<UsbVendorProductPair, UsbHardwareIdAndDriver> RegisteredHardware = new();

        static Hardware()
        {
            // .-------------.
            // | Stream Deck |
            // '-------------'

            var streamDeckKeys = new GridKeyLayout(5, 3, 72, 30);

            StreamDeck =
                RegisterNewHardwareInternal(
                    new
                    (
                        VendorIds.ElgatoSystemsGmbH,
                        ProductIds.StreamDeck
                    ),
                    "Stream Deck",
                    streamDeckKeys,
                    new HidComDriverStreamDeck()
                    {
                        // Limit of 3'200'000 bytes/s (~3.0 MiB/s)
                        // because without that limit glitches will happen on fast writes.
                        BytesPerSecondLimit = 3_200_000,
                    }
                );

            var streamDeckKeysNew = new GridKeyLayout(5, 3, 72, 32);

            StreamDeckMK2 =
                RegisterNewHardwareInternal(
                    new
                    (
                        VendorIds.ElgatoSystemsGmbH,
                        ProductIds.StreamDeckMK2
                    ),
                    "Stream Deck MK.2",
                    streamDeckKeysNew,
                    new HidComDriverStreamDeckJpeg(72)
                    {
                        // Limit of 1'500'000 bytes/s (~1.5 MB/s),
                        // because ImageGlitchTest.Rainbow has glitches with higher speeds
                        BytesPerSecondLimit = 1_500_000,
                    }
                );

            StreamDeckRev2 =
                RegisterNewHardwareInternal(
                    new
                    (
                        VendorIds.ElgatoSystemsGmbH,
                        ProductIds.StreamDeckRev2
                    ),
                    "Stream Deck Rev2",
                    streamDeckKeysNew,
                    new HidComDriverStreamDeckJpeg(72)
                    {
                        // Limit of 3'200'000 bytes/s (~3.0 MiB/s) just to be safe,
                        // because I don't own a StreamDeck Rev2 to test it.
                        BytesPerSecondLimit = 3_200_000,
                    }
                );

            // .----------------.
            // | Stream Deck XL |
            // '----------------'

            var streamDeckXlkeys = new GridKeyLayout(8, 4, 96, 38);
            var streamDeckXlDriver = new HidComDriverStreamDeckJpeg(96);

            StreamDeckXL =
                RegisterNewHardwareInternal(
                    new
                    (
                        VendorIds.ElgatoSystemsGmbH,
                        ProductIds.StreamDeckXL
                    ),
                    "Stream Deck XL",
                    streamDeckXlkeys,
                    streamDeckXlDriver
                );

            StreamDeckXlRev2 =
                RegisterNewHardwareInternal(
                    new
                    (
                        VendorIds.ElgatoSystemsGmbH,
                        ProductIds.StreamDeckXLRev2
                    ),
                    "Stream Deck XL Rev2",
                    streamDeckXlkeys,
                    streamDeckXlDriver
                );

            // .------------------.
            // | Stream Deck Mini |
            // '------------------'

            var streamDeckMiniKeys = new GridKeyLayout(3, 2, 80, 32);
            var streamDeckMiniDriver = new HidComDriverStreamDeckMini(80);

            StreamDeckMini =
                RegisterNewHardwareInternal(
                    new
                    (
                        VendorIds.ElgatoSystemsGmbH,
                        ProductIds.StreamDeckMini
                    ),
                    "Stream Deck Mini",
                    streamDeckMiniKeys,
                    streamDeckMiniDriver
                );

            SteamDeckMiniRev2 =
                RegisterNewHardwareInternal(
                    new
                    (
                        VendorIds.ElgatoSystemsGmbH,
                        ProductIds.StreamDeckMiniRev2
                    ),
                    "Stream Deck Mini Rev2",
                    streamDeckMiniKeys,
                    streamDeckMiniDriver
                );
        }

        /// <summary>
        /// Details about the classic Stream Deck
        /// </summary>
        public static IUsbHidHardware StreamDeck { get; }

        /// <summary>
        /// Details about the updated Stream Deck MK.2
        /// </summary>
        public static IUsbHidHardware StreamDeckMK2 { get; }

        /// <summary>
        /// Details about the classic Stream Deck Rev 2
        /// </summary>
        public static IUsbHidHardware StreamDeckRev2 { get; }

        /// <summary>
        /// Details about the Stream Deck XL
        /// </summary>
        public static IUsbHidHardware StreamDeckXL { get; }

        /// <summary>
        /// Details about the Stream Deck XL Rev2
        /// </summary>
        public static IUsbHidHardware StreamDeckXlRev2 { get; }

        /// <summary>
        /// Details about the Stream Deck Mini
        /// </summary>
        public static IUsbHidHardware StreamDeckMini { get; }

        /// <summary>
        /// Details about the Stream Deck Mini Rev2
        /// </summary>
        public static IUsbHidHardware SteamDeckMiniRev2 { get; }

        /// <summary>
        /// This method registers a new (currently unknown to this library) hardware driver.
        /// </summary>
        /// <remarks>
        /// <para>
        /// This method can be used if a new stream deck hardware is released to the market and
        /// the library currently doesn't have support for that new device. In the past a new device
        /// was often pretty similar to an existing device so with this method a tech-savvy person
        /// can register that new device.
        /// </para>
        /// <para>
        /// This feature is a bit "low-level", just take a look at the source code
        /// if you are not sure what to do.
        /// </para>
        /// </remarks>
        /// <param name="usbId">The USB vendor and product ID.</param>
        /// <param name="deviceName">A human readable name of the device.</param>
        /// <param name="keyLayout">The key layout of the device.</param>
        /// <param name="driver">The code that is used to communicate to the device.</param>
        /// <returns>
        /// Returns a description of the device that can be used to open that device with
        /// <see cref="StreamDeck.OpenDevice(IUsbHidHardware[])"/> or
        /// <see cref="StreamDeck.EnumerateDevices(IUsbHidHardware[])"/>.
        /// </returns>
        public static IUsbHidHardware RegisterNewHardware(
            UsbVendorProductPair usbId,
            string deviceName,
            GridKeyLayout keyLayout,
            IStreamDeckHidComDriver driver
        )
        {
            return RegisterNewHardwareInternal(usbId, deviceName, keyLayout, driver);
        }

        internal static UsbHardwareIdAndDriver RegisterNewHardwareInternal(
            UsbVendorProductPair usbId,
            string deviceName,
            GridKeyLayout keyLayout,
            IStreamDeckHidComDriver driver
        )
        {
            var internalReference = new UsbHardwareIdAndDriver(usbId, deviceName, keyLayout, driver);

            RegisteredHardware.AddOrUpdate(usbId, internalReference, (_, _) => internalReference);

            return internalReference;
        }

        internal static IEnumerable<UsbHardwareIdAndDriver> GetInternalStreamDeckHardwareInfos()
        {
            return RegisteredHardware.Values.ToList();
        }

        internal static UsbHardwareIdAndDriver GetInternalHardwareInfos(UsbVendorProductPair usbId)
        {
            if (RegisteredHardware.TryGetValue(usbId, out var hardwareInfo))
            {
                return hardwareInfo;
            }

            return null;
        }
    }
}
